### PR TITLE
magento/magento2#: Unit test for \Magento\Review\Observer\CatalogProductListCollectionAppendSummaryFieldsObserver

### PR DESCRIPTION
**Pull request by atwixfirster**
### Description (*)

PR adds a unit test which cover `\Magento\Review\Observer\CatalogProductListCollectionAppendSummaryFieldsObserver` class.

### Fixed Issues (if relevant)

1. magento/magento2#<issue_number>: Issue title

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Thank you!

### CODE REVIEW

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+use Magento\Store\Model\StoreManagerInterface;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Test class for \Magento\Review\Observer\CatalogProductListCollectionAppendSummaryFieldsObserver
+ */
+class CatalogProductListCollectionAppendSummaryFieldsObserverTest extends TestCase
+{
+    /**
+     * @var Event|\PHPUnit_Framework_MockObject_MockObject
```
Comment (lbajsarowicz) [inline app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: Actually with PHPUnit 6 you can use `\PHPUnit\Framework\MockObject\MockObject`

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+        $this->storeManagerMock
+            ->expects($this->once())
+            ->method('getStore')
+            ->willReturn($this->storeMock);
+
+        $this->storeMock
+            ->expects($this->once())
+            ->method('getId')
+            ->willReturn('1');
```
Comment (lbajsarowicz) [inline app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: Replace magic numbers with the corresponding const - for example `private const STUB_STORE_ID`

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            ->disableOriginalConstructor()
+            ->setMethods(['create'])
+            ->getMock();
+
+        $this->sumResourceMock = $this->getMockBuilder(Summary::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['appendSummaryFieldsToCollection'])
+            ->getMock();
+
+        $this->observer = new CatalogProductListCollectionAppendSummaryFieldsObserver(
```
Comment (lbajsarowicz) [inline app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php:106]: Use ObjectManager helper: `\Magento\Framework\TestFramework\Unit\Helper\ObjectManager`

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 2 above)
Comment (lbajsarowicz) [inline app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: You test the implementation, without covering the most important part that is the goal of Observer: 
```php
...->appendSummaryFieldsToCollection(
            $productCollection,
            $this->storeManager->getStore()->getId(),
            \Magento\Review\Model\Review::ENTITY_PRODUCT_CODE
        )
```

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Cover observer's most important part with Unit Test.

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 2 above)
Comment (atwixfirster) [reply app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: **done**

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 3 above)
Comment (atwixfirster) [reply app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php:106]: Proposition with `ObjectManager helper` - does not work properly.

I've used the next `$this->observer` initialization:

```
        $this->observer = $this->objectManagerHelper->getObject(
            CatalogProductListCollectionAppendSummaryFieldsObserver::class,
            [
                $this->sumResourceFactoryMock,
                $this->storeManagerMock
            ]
        );
```

And
 
```
$this->sumResourceFactory->create()
```

returns a NULL in the `Magento\Review\Observer\CatalogProductListCollectionAppendSummaryFieldsObserver`::`execute()`.

When I use 

```
        $this->observer = new CatalogProductListCollectionAppendSummaryFieldsObserver(
                $this->sumResourceFactoryMock,
                $this->storeManagerMock
        );
```

then 

```
$this->sumResourceFactory->create()
```

returns a mock of `Magento\Review\Model\ResourceModel\Review\SummaryFactory` mock object.

There are a lot of unit tests which uses a direct Observer class initialization:
- Magento\Captcha\Test\Unit\Observer\CheckUserLoginBackendObserverTest
- Magento\CatalogUrlRewrite\Test\Unit\Observer\CategoryProcessUrlRewriteMovingObserverTest
- Magento\CurrencySymbol\Test\Unit\Observer\CurrencyDisplayOptionsTest
- Magento\Customer\Test\Unit\Model\ResourceModel\Customer\GridTest
- ...

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 1 above)
Comment (atwixfirster) [reply app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: **done**

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 2 above)
Comment (atwixfirster) [reply app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: **done**

**Comment by atwixfirster** [comment]
@lbajsarowicz, all your suggestions have been implemented except a proposition with `ObjectManager helper`. 

Please review. 

Thanks

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    /**
+     * Testable Object
+     *
+     * @var CatalogProductListCollectionAppendSummaryFieldsObserver
+     */
+    private $observer;
+
+    /**
+     * @var Observer|\PHPUnit\Framework\MockObject\MockObject
```
Comment (lbajsarowicz) [inline app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: Import class :-)

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->sumResourceFactoryMock
+            ->expects($this->once())
+            ->method('create')
+            ->willReturn($this->sumResourceMock);
+
+        $this->sumResourceMock
+            ->expects($this->once())
+            ->method('appendSummaryFieldsToCollection')
+            ->willReturn($this->sumResourceMock);
```
Comment (lbajsarowicz) [inline app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php:145]: :+1:

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 3 above)
Comment (lbajsarowicz) [reply app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php:106]: OK. I'll investigate that further later on. But for now, it's fine for me :+1:

**Review by lbajsarowicz — APPROVED** [review]
Great job.
Just import the MockObject class and we're done : )

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6591](https://jira.corp.magento.com/browse/ENGCOM-6591) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php (hunk 4 above)
Comment (atwixfirster) [reply app/code/Magento/Review/Test/Unit/Observer/CatalogProductListCollectionAppendSummaryFieldsObserverTest.php]: **done**

**Comment by atwixfirster** [comment]
> Great job.

thanks ;)

> Just import the MockObject class and we're done : )

**done**

**Review by dmytro-ch — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch, thank you for the review.
[ENGCOM-6591](https://jira.corp.magento.com/browse/ENGCOM-6591) has been created to process this Pull Request

**Review by dmytro-ch — CHANGES_REQUESTED** [review]
@atwixfirster, could you please check and fix the static test failures?

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6591](https://jira.corp.magento.com/browse/ENGCOM-6591) has been created to process this Pull Request

**Review by dmytro-ch — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch, thank you for the review.
[ENGCOM-6591](https://jira.corp.magento.com/browse/ENGCOM-6591) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**